File: Java/Homework04/Homework04_02/Homework04_02.java
/*
 * File name : Homework04_02.java
 * Author : Kim Yeong Ho
 * Date : July 13, 2023
 * Major feature : 
 * - 다음 그림과 같은 구조를 가지는  double 타입의 2차원 배열을 선언하고,
 * - 값을 직접 배열에 삽입하는 방법으로 그림에 있는 값으로 초기화하고, 출력하는 코드를 작성하시오.
 * - 1.1  1.2  1.3  1.4
 * - 2.1  2.2
 * - 3.1  3.2  3.3
 * - 4.1
 */
package Homework04_02;

public class Homework04_02 {
	public static void main(String[] args) {
		final int row = 4;
		final int[] col = {4, 2, 3, 1};
		
		double arr[][] = new double[row][];
		for (int i = 0; i < arr.length; i++) {
			arr[i] = new double[col[i]];
			for (int j = 0; j <arr[i].length; j++) {
				arr[i][j] = (i + 1) + (j + 1) * 0.1;
			}
		}
		
		for (int i = 0; i < arr.length; i++) {
			for (int j = 0; j <arr[i].length; j++) {
				System.out.print(arr[i][j] + "\t");
			}
			System.out.println();
		}
	}
}

File: C/Homework04_02/Homework04_02.c
/*
* File name : Homework04_02.c
* Author : Kim Yeong Ho
* Date : July 26, 2023
* Major feature :
* - 사용자로부터 색깔의 종류의 개수를 입력 받아, 주어진 사각형의 4가지 영역을
*   아래의 조건을 만족하면서 채우는 모든 방법과 전체 가짓수를 찾아내는 프로그램을 작성하시오.
*   - 색깔의 종류의 개수는 2, 3, 4로 한정
*     - 색깔 2개: Red, Green
*     - 색깔 3개: Red, Green, Blue
*     - 색깔 4개: Red, Green, Blue, Yellow
*   - 인접한 영역은 동일한 색깔을 가질 수 없음
*   - 조건을 만족하는 4가지 영역 coloring 예시:
*     (1) : Red		(2) : Green
*     (3) : Blue	(4) : Yellow
*/
#include <stdio.h>

int count_color(int a, int b, int c, int d);
void print_color(int num, int order);

int main(void) {
	int num, order, index = 0, count = 0;

	printf(" - number of colors (2 ~ 4): ");
	scanf("%d", &num);

	for (int a = 0; a < num; a++) {
		for (int b = 0; b < num; b++) {
			for (int c = 0; c < num; c++) {
				for (int d = 0; d < num; d++) {
					if (a != b && a != c && b != d && c != d) {
						order = 1;	count++;
						printf(" %4d:", count);
						print_color(a, order++);
						print_color(b, order++);
						print_color(c, order++);
						print_color(d, order);
						printf(" # of colors = %d", count_color(a, b, c, d));
						printf("\n");
					}
				}
			}
		}
	}
	printf(" - Total number of coloring: %d\n", count);

	return 0;
}

int count_color(int a, int b, int c, int d) {
	int index = 0;

	for (int i = 0; i < 4; i++)
		if (a == i || b == i || c == i || d == i)
			index++;

	return index;
}

void print_color(int num, int order) {
	switch (num) {
		case 0: printf(" (%d):    Red ", order);	break;
		case 1: printf(" (%d):  Green ", order);	break;
		case 2: printf(" (%d):   Blue ", order);	break;
		case 3: printf(" (%d): Yellow ", order);	break;
		default: printf(" (%d): Error ", order);	break;
	}
}